File: algorithms/sorting/quick-sort/QuickSort.c
int partition(int arr[], int start, int end) {
  int pivot = arr[end];
  int i = start - 1;

  for (int j = start; j < end; j++) {
    if (arr[j] < pivot) {
      i++;
      int temp = arr[i];
      arr[i] = arr[j];
      arr[j] = temp;
    }
  }

  i++;
  int temp = arr[i];
  arr[i] = arr[end];
  arr[end] = temp;

  return i;
}

void quick_sort(int arr[], int start, int end) {
  if (start >= end)
    return;

  int pivot = partition(arr, start, end);

  quick_sort(arr, start, pivot - 1);
  quick_sort(arr, pivot + 1, end);
}

File: algorithms/sorting/quick-sort/QuickSort.dart
import 'dart:math';

class Sorting {
  static List<int> QuickSort(List<int> unsortedList, int low, int high) {
    if (low < high) {
      int pivotIndex = Sorting._Partition(unsortedList, low, high);
      Sorting.QuickSort(unsortedList, low, pivotIndex - 1);
      Sorting.QuickSort(unsortedList, pivotIndex + 1, high);
    }

    return unsortedList;
  }

  static int _Partition(List<int> unsortedList, int low, int high) {
    int pivot = unsortedList[high];
    int i = low - 1;
    int temp;

    for (int j = low; j < high; j++) {
      if (unsortedList[j] < pivot) {
        i++;
        temp = unsortedList[i];
        unsortedList[i] = unsortedList[j];
        unsortedList[j] = temp;
      }
    }

    temp = unsortedList[i + 1];
    unsortedList[i + 1] = unsortedList[high];
    unsortedList[high] = temp;
    return i + 1;
  }
}
